From e344c697824d8e515083595caf72c74b16d74d8a Mon Sep 17 00:00:00 2001
From: Yiwei Zhang <zzyiwei@chromium.org>
Date: Tue, 20 Oct 2026 17:02:37 -0700
Subject: [PATCH 3/4] server: fix a memcpy UB when no fds are received

Unlike the proxy side, server side doesn't know which context op is
arriving before receving the message. So for those cmds without fds, we
should skip memcpy to avoid UB.

Part-of: <https://gitlab.freedesktop.org/virgl/virglrenderer/-/merge_requests/1698>
(cherry picked from commit 06abc4f68cf02dbd0fae9b79ba22b83c1a356f6f)
---
 server/render_socket.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/server/render_socket.c b/server/render_socket.c
index 58edcf46..80d3a10a 100644
--- a/server/render_socket.c
+++ b/server/render_socket.c
@@ -145,7 +145,8 @@ render_socket_receive_request_internal(struct render_socket *socket,
       const int *received_fds = get_received_fds(&msg, &received_fd_count);
       assert(received_fd_count <= max_fd_count);
 
-      memcpy(fds, received_fds, sizeof(*fds) * received_fd_count);
+      if (received_fd_count)
+         memcpy(fds, received_fds, sizeof(*fds) * received_fd_count);
       *out_fd_count = received_fd_count;
    } else if (out_fd_count) {
       *out_fd_count = 0;
-- 
2.43.0

